src/components/flowchart/flowchart.js

```diff
@@ -528,34 +538,80 @@ export class FlowChart extends Component {
     // then on a single node click, it should reset the sliced pipeline state
     if (from !== null && to !== null) {
       this.updateSlicedPipelineState(null, null, []);
+      this.setState({ showSlicingNotification: false }); // Hide notification
     } else {
       // Else, set the first node as the 'from' node based on current state
       // we need this so that if user hold shift and click on a second node,
       // the 'from' node is already set
       this.updateSlicedPipelineState(id, to, range);
+      this.setState({ showSlicingNotification: true }); // Show notification
     }
 
     // Reset the slicePipeline on single node click
     this.props.onSlicePipeline(null, null);
   };
 
+  /**
+   * Determines the correct order of nodes based on their positions,
+   * but keeps the user's original selection as the visual 'from' node.
+   * @param {string} userSelectedFromNodeId - User's initially selected 'from' node ID.
+   * @param {string} toNodeId - The current 'to' node ID.
+   * @returns {Object} - Object containing userVisualFromNodeId and adjustedFromNodeId, newToNodeId
+   */
+  determineNodesOrder = (userSelectedFromNodeId, toNodeId) => {
+    // Get bounding client rects of nodes
+    const fromNodeElement = document.querySelector(
+      `[data-id="${userSelectedFromNodeId}"]`
+    );
+    const toNodeElement = document.querySelector(`[data-id="${toNodeId}"]`);
+
+    if (!fromNodeElement || !toNodeElement) {
+      return {
+        userVisualFromNodeId: null,
+        adjustedFromNodeId: null,
+        newToNodeId: null,
+      }; // If any element is missing, return nulls
+    }
+
+    const fromNodeRect = fromNodeElement.getBoundingClientRect();
+    const toNodeRect = toNodeElement.getBoundingClientRect();
+
+    // Reorder the nodes based on their Y-coordinate
+    const [adjustedFromNodeId, newToNodeId] =
+      fromNodeRect.y < toNodeRect.y
+        ? [userSelectedFromNodeId, toNodeId] // Keep order
+        : [toNodeId, userSelectedFromNodeId]; // Swap if needed
+
+    return {
+      userVisualFromNodeId: userSelectedFromNodeId, // Keep user's selection visually as 'from'
+      adjustedFromNodeId,
+      newToNodeId,
+    };
+  };
+
   handleShiftClick = (node) => {
     // Close meta data panel
     this.props.onLoadNodeData(null);
+    const { from: userSelectedFromNodeId, range } =
+      this.state.slicedPipelineState;
 
-    const fromNodeId = this.state.slicedPipelineState.from || node.id;
+    const fromNodeId = userSelectedFromNodeId || node.id; // Keep existing 'from' node or set the current one if not set
```

Comment:
I think the simpler way will be:
- whatever user choses, we update the state, because the state is simply just for changing the background colour of the element so it doesn't really matter if the from node is from the bottom
- we only change the order when we update the redux store. 

So something like this will be better I think:

```
const fromNodeId = fromNodeIdState || node.id;
const toNodeId = node.id;

// update the state from what user has clicked
this.updateSlicedPipelineState(fromNodeId, toNodeId, range);

const { updatedFromNodeId, updatedToNodeId } = this.determineNodesOrder(
      fromNodeId,
      toNodeId
 );

 // Slice the pipeline based on the determined node order
 // If the order could not be determined, use the original selection
 if (updatedFromNodeId && updatedToNodeId) {
      this.props.onSlicePipeline(updatedFromNodeId, updatedToNodeId);
  } else {
      this.props.onSlicePipeline(fromNodeId, toNodeId);
  }

this.props.onApplySlice(false);
this.setState({ showSlicingNotification: false }); // Hide notification after selecting the second node

```

Reply:
this way, `determineNodesOrder` only returns 2 updated values, rather than 3

Reply:
Done, this has simplified it!